### PR TITLE
s3 api bucket policy implementaion

### PR DESCRIPTION
**Pull request by manuvakery1**
Issue:
bucket policy is not implemented in the s3api middleware

Fix:
1. added APIS required for bucket policy 
2. updated core APIs to make use of bucket policy for access control

### CODE REVIEW

File: swift/common/middleware/s3api/acl_handlers.py [hunk 1]
```diff
@@ -140,7 +144,12 @@ def _handle_acl(self, app, sw_method, container=None, obj=None,
         elif resource == 'container':
             resp = self.req.get_acl_response(app, 'HEAD',
                                              container, '')
+            self.logger.debug("resp %s", resp)
             acl = resp.bucket_acl
+            self.logger.debug("setting bucket policy in response")
+            if hasattr(resp, 'bucket_policy'):
+                policy = resp.bucket_policy
+                #TODO: use policy to vallidate ACL
```
Comment (diganta-dutta) [inline swift/common/middleware/s3api/acl_handlers.py]: Why is this required? The bucket policy related functionality should be handled by the derived class only.

File: swift/common/middleware/s3api/acl_handlers.py [hunk 2]
```diff
@@ -181,7 +190,124 @@ def get_acl(self, headers, body, bucket_owner, object_owner=None):
...
+
+    def __init__(self, req, logger, container=None, obj=None, headers=None):
+        super(BucketPolicyHandler, self).__init__(req, logger, container, obj, headers)
+
+    def get_bucket_policy(self, body):
+        """
+        Get BucketPolicy instance from json body.
+        """
+        self.logger.debug("body %s", body)
+        self.logger.debug("body %s", type(body))
```
Comment (diganta-dutta) [inline swift/common/middleware/s3api/acl_handlers.py]: Combine multiple logger calls to a single.

File: swift/common/middleware/s3api/acl_handlers.py [hunk 3]
```diff
@@ -181,7 +190,124 @@ def get_acl(self, headers, body, bucket_owner, object_owner=None):
...
+        try:
+            import json
```
Comment (diganta-dutta) [inline swift/common/middleware/s3api/acl_handlers.py]: Move to global context

File: swift/common/middleware/s3api/acl_handlers.py [hunk 4]
```diff
@@ -181,7 +190,124 @@ def get_acl(self, headers, body, bucket_owner, object_owner=None):
...
+        This method expects to call Request._get_response() in outside of
+        this method so that this method returns response only when sw_method
+        is HEAD.
+        """
+        container = self.container if container is None else container
+        obj = self.obj if obj is None else obj
+        self.logger.debug("container %s", self.container)
+        self.logger.debug("obj %s", self.obj)
+        sw_method = sw_method or self.req.environ['REQUEST_METHOD']
+        resource = 'object' if obj else 'container'
```
Comment (diganta-dutta) [inline swift/common/middleware/s3api/acl_handlers.py:220]:
```suggestion
        resource = 'object' if obj is not None else 'container'
```
Otherwise `obj` can also be `''` which probably isn't correct. I think the above checks should be cleaned up. First decide if you are expecting any input parameters to have blank string. `None` and blank strings are not same.

File: swift/common/middleware/s3api/acl_handlers.py [hunk 5]
```diff
@@ -181,7 +190,124 @@ def get_acl(self, headers, body, bucket_owner, object_owner=None):
...
+        headers = self.headers if headers is None else headers
+
+        self.logger.debug(
+            'checking permission: %s %s %s %s' %
+            (container, obj, sw_method, dict(headers)))
+
+        if not container:
```
Comment (diganta-dutta) [inline swift/common/middleware/s3api/acl_handlers.py:227]: This will fails with blank container name also. Also, there should be an error logs probably. This is an exceptional situation.

File: swift/common/middleware/s3api/acl_handlers.py [hunk 6]
```diff
@@ -181,7 +190,124 @@ def get_acl(self, headers, body, bucket_owner, object_owner=None):
...
+            return
+
+        # self.logger.debug("resource %s", resource)
+
+        if not permission and (self.method, sw_method, resource) in ACL_MAP:
```
Comment (diganta-dutta) [inline swift/common/middleware/s3api/acl_handlers.py:232]: Lets discuss this ACL_MAP.

File: swift/common/middleware/s3api/acl_handlers.py [hunk 7]
```diff
@@ -181,7 +190,124 @@ def get_acl(self, headers, body, bucket_owner, object_owner=None):
...
+                    if version_id is None:
+                        query = {}
+                    else:
+                        query = {'version-id': version_id}
+                    resp = self.req.get_acl_response(app, 'HEAD',
+                                                     container, obj,
+                                                     headers, query=query)
+                    acl = resp.object_acl
+                acl.check_permission(self.user_id, permission)
+        except Exception as e:
```
Comment (diganta-dutta) [inline swift/common/middleware/s3api/acl_handlers.py:277]: Very generic. What errors are you expecting here?

File: swift/common/middleware/s3api/acl_handlers.py [hunk 8]
```diff
@@ -292,7 +418,32 @@ def POST(self, app):
             self._handle_acl(app, self.method)
 
 
-class MultiObjectDeleteAclHandler(BaseAclHandler):
+class S3BucketPolicyHandler(BucketPolicyHandler):
+
```
Comment (diganta-dutta) [inline swift/common/middleware/s3api/acl_handlers.py:398]: Please add docstring

File: swift/common/middleware/s3api/controllers/s3_bucket_policy.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    @public
+    def GET(self, req):
+        """
+        Handles GET Bucket Policy.
+        """
+        resp = req.get_response(self.app)
+        if resp.bucket_policy:
+            bucket_policy = resp.bucket_policy
+            resp = HTTPOk()
+            resp.body = json.dumps(bucket_policy.to_dict())
```
Comment (diganta-dutta) [inline swift/common/middleware/s3api/controllers/s3_bucket_policy.py:42]: exception handling required.

File: swift/common/middleware/s3api/subresource.py [hunk 10]
```diff
@@ -568,10 +601,438 @@ def acl(key, bucket_owner, object_owner=None,
...
+def from_str(x):
+    return x
+
+
+def to_class(c, x):
+    return x.to_dict()
+
+
+
+class IPAddress:
```
Comment (diganta-dutta) [inline swift/common/middleware/s3api/subresource.py]: Please add docstring explaining these Classes.

File: test/functional/s3api/test_bucket_policy.py [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+            "Version": "2020-04-06",
+            "Statement": [
+                {
+                    "Effect": "Allow",
+                    "Principal": {
+                        "AWS": "*"
+                    },
+                    "Action": "s3:GetObject",
+                    "Resource": [
+                        "arn:aws:s3:::bp-root/*"
```
Comment (diganta-dutta) [inline test/functional/s3api/test_bucket_policy.py:72]: what is `bp-root`?

File: test/functional/s3api/test_bucket_policy.py [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+        status, headers, body = \
+            self.conn3.make_request('PUT', self.bucket, query='policy')
+        self.assertEqual(get_error_code(body), 'AccessDenied')
+
+        self.conn.make_request('PUT', self.bucket, None)
+        malformed_policy = {
+            "Version": "2020-04-06",
+        }
+        status, headers, body = \
+            self.conn.make_request('PUT', self.bucket, body=json.dumps(malformed_policy),
```
Comment (diganta-dutta) [inline test/functional/s3api/test_bucket_policy.py]: Please add a test with body which is not json.

File: test/functional/s3api/test_bucket_policy.py [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+        if 's3_access_key3' not in tf.config or \
+                's3_secret_key3' not in tf.config:
+            raise tf.SkipTest(
+                'TestS3Acl requires s3_access_key3 and s3_secret_key3 '
+                'configured for reduced-access user')
+        status, headers, body = self.conn.make_request('PUT', self.bucket)
+        self.assertEqual(status, 200, body)
+        access_key3 = tf.config['s3_access_key3']
+        secret_key3 = tf.config['s3_secret_key3']
+        self.conn3 = Connection(access_key3, secret_key3, access_key3)
```
Comment (diganta-dutta) [inline test/functional/s3api/test_bucket_policy.py:48]: Please add comments for what `conn2` and `conn3` are used.

File: test/functional/s3api/test_bucket_policy.py [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+        self.assertEqual(
+            ctx.exception.response['ResponseMetadata']['HTTPStatusCode'], 403)
+        self.assertEqual(ctx.exception.response['Error']['Code'], 'AccessDenied')
+        policy_dict = {
+            "Version": "2020-04-06",
+            "Statement": [
+                {
+                    "Effect": "Allow",
+                    "Principal": {
+                        "AWS": "*"
```
Comment (diganta-dutta) [inline test/functional/s3api/test_bucket_policy.py:191]: What about testing with specific principal of the user provided access.

File: test/functional/s3api/test_bucket_policy.py [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+        status, headers, body = \
+            self.conn.make_request('PUT', self.bucket, body=json.dumps(policy_dict),
+                                   query="policy")
+        self.assertEqual(status, 200)
+        status, headers, body = \
+            self.conn2.make_request('DELETE', self.bucket, obj)
+        self.assertEqual(status, 204)
+
+
+class TestS3AclSigV4(TestS3BucketPolicy):
```
Comment (diganta-dutta) [inline test/functional/s3api/test_bucket_policy.py:325]: Tests missing?

File: swift/common/middleware/s3api/acl_handlers.py (hunk 4 above)
Comment (manuvakery1) [reply swift/common/middleware/s3api/acl_handlers.py:220]: we already have check for none `obj = self.obj if obj is None else obj` . We are good here

File: test/functional/s3api/test_bucket_policy.py (hunk 15 above)
Comment (manuvakery1) [reply test/functional/s3api/test_bucket_policy.py:325]: no . the tests are the same for SigV4 too except for setting the environment

File: test/functional/s3api/test_bucket_policy.py (hunk 11 above)
Comment (manuvakery1) [reply test/functional/s3api/test_bucket_policy.py:72]: the sample input is taken from aws docs only for testing. The actual policy will differ in idrivecloud

File: test/functional/s3api/test_bucket_policy.py (hunk 14 above)
Comment (manuvakery1) [reply test/functional/s3api/test_bucket_policy.py:191]: https://github.com/idrive-online-backup/swift/blob/feature/bucket-policy/test/functional/s3api/test_bucket_policy.py#L203  This test covers that
